### ISSUE TITLE
Comparison to torchmetrics

### ISSUE THREAD

**Issue by rsokl** [pytorch/torcheval#82]
Hello! `torcheval` looks great!

I'd be interested to know how `torcheval` compares to [torchmetrics](https://github.com/Lightning-AI/metrics). Are there certain shortcomings in `torchmetrics` that `torcheval` hopes to address? Any other insights into what inspired the creation of `torcheval` might help users understand what makes this project unique 😄

**Comment by yongen9696** [comment]
> what makes this project unique
I wonder as well, [torchmetrics](https://github.com/Lightning-AI/metrics) is quite mature and complete with available metric tools.

**Comment by ninginthecloud** [comment]
Hi, @rsokl and @yongen9696 ,  thanks for the great question~ 

## Kudos to the community
Support for metrics and evaluation has been a long-running [request](https://github.com/pytorch/pytorch/issues/22439) from the PyTorch community. First, we would like to give kudos to [scikit-learn metrics](https://scikit-learn.org/stable/modules/classes.html#sklearn-metrics-metrics), [Keras Metrics](https://keras.io/api/metrics/), [Ignite Metrics](https://pytorch.org/ignite/metrics.html), and [TorchMetrics](https://torchmetrics.readthedocs.io/en/stable/) as existing projects in the ML community that have inspired TorchEval. In particular, we have previously discussed these [design](https://github.com/Lightning-AI/metrics/issues/344) points with the developers of TorchMetrics.

## What makes TochEval unique? 
### Philosophy for TorchEval
TorchEval is a library that enables easy and performant model evaluation for PyTorch. The library’s philosophy is to provide minimal interfaces that are bolstered by a robust toolkit, alongside a rich collection of out-of-the-box performant implementations. Critically, we believe in the following axes:

- No surprises in behavior
- Fast by default
- Easily extensible
- Works naturally in distributed applications

### Compared to TorchMetrics
These are the major differences between TorchEval and TorchMetrics:

**Metric synchronization in distributed applications**

1. Metric synchronization is supported through the [toolkit](https://github.com/pytorch/torcheval/blob/main/torcheval/metrics/toolkit.py), not on the base interface.
2. Metric synchronization does not change metric states in-place, which means users don’t need to worry about undefined transitions (e.g. calling `update()` after `sync()`) or rewinding to previous states.
3. Metric synchronization must be explicitly opted into by users. This makes it easy for callers to distinguish between results on a particular rank vs global results. Unlike TorchMetrics, there is no default synchronization on step, which has a significant performance overhead in distributed applications.
4. **Explicit**: Synchronizing states in TorchEval operate on the whole metric object, not values per state. Specifically, TorchEval requires users to implement a `merge_state()` method to define how to synchronize the states. This avoids assumptions of the state object compared to `dist_reduce_fx` in TorchMetrics.
5. **Performance**: When synchronizing a metric, the TorchEval toolkit runs only 1 collective per metric. In TorchMetrics, there’s 1 collective per state per metric, which requires more resources.
6. **Extensible**: The toolkit for metric synchronization today covers the typical [SPMD](https://en.wikipedia.org/wiki/Single_program,_multiple_data) use case, but can be extended to cover peer-to-peer use cases (e.g. via torch.distributed.rpc) without changing the base interface.

**Interface clarity**

1. Class-based metrics in TorchEval offer only `update`, `compute`, `reset`, and `merge_state` methods, which makes it obvious to callers what states are used for computing results. There’s only **1** way to get results from class-based metrics. This differs from TorchMetrics, where metrics are also `nn.Module`s that have `forward` defined. As a result, users have **2** choices of how to update their metrics: call `forward` **or** `update`. `forward` from TorchMetrics is much more involved and can be an inadvertent system performance drain.
2. In the base `Metric` interface, TorchEval does not wrap the `update()` or `compute()` methods implemented by callers. Compared to users implementing a metric with TorchMetrics, there are no surprises in TorchEval. 

**Performance**

1. We believe the biggest performance benefits come from the explicit interfaces offered. In addition to the points listed above on synchronization, out-of-the-box implementations offered are optimized with:
     1. Vectorization ([example](https://github.com/pytorch/torcheval/blob/c9ce95f1b3fb973ec39014065214ed042f36e495/torcheval/metrics/functional/classification/f1_score.py#L182))
     2. JIT scripting ([example](https://github.com/pytorch/torcheval/blob/main/torcheval/metrics/functional/classification/auroc.py#L112))
     3. Binned metrics ([example](https://github.com/pytorch/torcheval/blob/main/torcheval/metrics/classification/binned_auroc.py))
2. Looking forward, TorchEval is exploring custom kernels and/or Triton integrations to further accelerate computation.

### Beyond Metrics
TorchEval also includes tools for evaluation like [FLOPs](https://github.com/pytorch/torcheval/blob/main/torcheval/tools/flops.py) and summarization techniques for [modules](https://github.com/pytorch/torcheval/blob/main/torcheval/tools/module_summary.py).

cc: @ananthsub @bobakfb @JKSenthil

**Comment by ananthsub** [comment]
I think @ninginthecloud 's reply summaries the difference very well, so I'll close out this issue. @rsokl please let us know if you have further questions about this though!

**Comment by rsokl** [comment]
Thank you! This response was very useful.

**Comment by rsokl** [comment]
(given the engagement on this thread, you might consider pinning it in your issues section so that other inquiring users can find it easily 😄 )

**Comment by williamFalcon** [comment]
Hi! William here from Lightning. The Lightning team led the development of torchmetrics. There was a period where @ananthsub was a close member of the torchmetrics team where the impression that we were under was that he was contributing back to Lightning Torchmetrics OSS. 

However, it seems that when Lightning did not agree to give torchmetrics to Meta, the team decided to do their own thing which is in their full right. 

I’m sure that eval is a good attempt at metrics and you can be the judge of what you prefer to use. What I can say is that we have a whole company dedicated to making sure our software is the best in the world and are committed to providing first class support and iterating all the feedback into torchmetrics. 

Thanks for the thorough comparison! we will be taking this feedback into consideration as we prepare for our next release. 

cheers!
